src/github3/repos/branch.py

```diff
@@ -653,6 +660,19 @@ def users(self, number=-1):
 
         return self._iter(int(number), self.users_url, users.ShortUser)
 
+    def as_update_dict(self):
+        curr_dict = self.as_dict()
```

Comment:
Why can't we make `as_dict()` smarter in general so this doesn't bite any other endpoint usage?

Reply:
Looking at this now, and thinking about it again, I don't think it's the "as_dict()" behaviour that I want to change.

Calling the update function on the BranchProtection object is where the error is thrown, and that occurs once we make the API call.

Getting back full information from the as_dict() calls on the ProtectionRestrictions and ProtectionRequiredPullRequestReviews objects is good for general use in code, and I was attempting not to affect that.  Using the return from said actions in an update on the BranchProtection after modification to the local objects is probably desirable behaviour?

Therefore, if the update function made the appropriate change to the edit variable prior to the API call, we would be validating both function inputs and also current state returns without modifying the child objects.

Does this seem like a better change to you?